src/tools/tidy/src/error_codes.rs

```diff
@@ -0,0 +1,397 @@
+        if !has_test.1 && !test_ignored {
+            errors.push(format!(
+                "`{}` doesn't use its own error code in compile_fail example",
+                path.display(),
+            ));
+        } else if has_test.1 && test_ignored {
+            errors.push(format!(
+                "`{}` has a compile_fail doctest with its own error code, it shouldn't \
+                be listed in `IGNORE_DOCTEST_CHECK`",
+                path.display(),
+            ));
+        }
+    });
+
+    if emit_ignore_warning > 0 {
+        println!(
+            "WARNING: {emit_ignore_warning} error codes use the ignore header. This should not be used, add the error codes to the \
+            `IGNORE_DOCTEST_CHECK` constant instead. This *will* become a hard error."
+        );
+    }
+    if emit_no_code_warning > 0 {
+        println!(
+            "WARNING: {emit_ignore_warning} error codes don't have a code example, all error codes are expected \
+            to have one (even if untested). This *will* become a hard error."
+        );
+    }
+    if emit_no_longer_warning > 0 {
+        println!(
+            "WARNING: {emit_no_longer_warning} error codes are no longer emitted and should be removed entirely. \
+            This *will* become a hard error."
+        );
+    }
+
+    no_longer_emitted_codes
+}
+
+/// This function returns a tuple indicating whether the provided explanation:
+/// a) has a code example, tested or not.
+/// b) has a valid doctest
+fn check_explanation_has_doctest(explanation: &str, err_code: &str) -> (bool, bool, bool, bool) {
+    let mut found_code_example = false;
+    let mut found_proper_doctest = false;
+
+    let mut emit_ignore_warning = false;
+    let mut emit_no_longer_warning = false;
+
+    for line in explanation.lines() {
+        let line = line.trim();
+
+        if line.starts_with("```") {
+            found_code_example = true;
+
+            // Check for the `rustdoc` doctest headers.
+            if line.contains("compile_fail") && line.contains(err_code) {
+                found_proper_doctest = true;
+            }
+
+            if line.contains("ignore") {
+                emit_ignore_warning = true;
+                found_proper_doctest = true;
+            }
+        } else if line
+            .starts_with("#### Note: this error code is no longer emitted by the compiler")
+        {
+            emit_no_longer_warning = true;
+            found_code_example = true;
+            found_proper_doctest = true;
+        }
+    }
+
+    (found_code_example, found_proper_doctest, emit_ignore_warning, emit_no_longer_warning)
+}
+
+// Stage 3: Checks that each error code has a UI test in the correct directory
+fn check_error_codes_tests(root_path: &Path, error_codes: &[String], errors: &mut Vec<String>) {
+    let tests_path = root_path.join(Path::new(ERROR_TESTS_PATH));
+
+    // Some warning counters, this whole thing is clunky but'll be removed eventually.
+    let mut no_ui_test = 0;
```

Comment:
Instead of a counter, why not have a vec of strings like you do elsewhere, and just print them if they're not empty? Otherwise this is just going to print that a bunch of these codes need tests, but not *which ones*.

Reply:
I did this originally, and I agree that it's simpler. However with the new warnings I've introduced, there are ~250 warnings that would be emitted. The warnings aren't really permanent and are more just tracking progress. When these are fixed (I'll have some follow up PRs), they'll become hard errors, which simplifies the code a lot.